src/conn_sock.c

```diff
@@ -111,51 +145,150 @@ char *setup_attach_socket(void)
+		lcl_notify_host_fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
+		if (lcl_notify_host_fd == -1) {
+			pexit("Failed to create notify socket");
+		}
+		lcl_notify_host_addr.sun_family = AF_UNIX;
+		strncpy(lcl_notify_host_addr.sun_path, socket_path, sizeof(lcl_notify_host_addr.sun_path) - 1);
+	}
+
+	/*
+	 * Create a symlink so we don't exceed unix domain socket
+	 * path length limit.
+	 */
+	_cleanup_free_ char *notify_symlink_dir_path = g_build_filename(opt_socket_path, opt_cuuid, NULL);
+	if (unlink(notify_symlink_dir_path) == -1 && errno != ENOENT)
+		pexit("Failed to remove existing symlink for notify socket directory");
+
+	/*
+	 * This is to address a corner case where the symlink path length can end up being
+	 * the same as the socket.  When it happens, the symlink prevents the socket from being
+	 * be created.  This could still be a problem with other containers, but it is safe
+	 * to assume the CUUIDs don't change length in the same directory.  As a workaround,
+	 *  in such case, make the symlink one char shorter.
+	 */
+	if (strlen(notify_symlink_dir_path) == (sizeof(notify_addr.sun_path) - 1))
+		notify_symlink_dir_path[sizeof(notify_addr.sun_path) - 2] = '\0';
+
+	if (symlink(opt_bundle_path, notify_symlink_dir_path) == -1)
+		pexit("Failed to create symlink for notify socket");
+
+	_cleanup_free_ char *notify_sock_fullpath = g_build_filename(opt_socket_path, opt_cuuid, "notify/notify.sock", NULL);
+	_cleanup_free_ char *notify_sock_path = g_build_filename(opt_cuuid, "notify/notify.sock", NULL);
+	ninfof("notify sock path: %s", notify_sock_fullpath);
+
+	strncpy(notify_addr.sun_path, notify_sock_path, sizeof(notify_addr.sun_path) - 1);
+	ninfof("addr{sun_family=AF_UNIX, sun_path=%s}", notify_addr.sun_path);
+
+	/*
+	 * We make the socket non-blocking to avoid a race where client aborts connection
+	 * before the server gets a chance to call accept. In that scenario, the server
+	 * accept blocks till a new client connection comes in.
+	 */
+	if (getcwd(cwd, sizeof(cwd)) == NULL)
+		pexit("Failed to get CWD for notify socket");
+
+	notify_socket_fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
+	if (notify_socket_fd == -1)
+		pexit("Failed to create notify socket");
+
+	rmt_notify_sock.fd = notify_socket_fd;
+
+	if (fchmod(notify_socket_fd, 0777))
+		pexit("Failed to change notify socket permissions");
+
+	if (chdir(opt_socket_path) == -1)
+		pexitf("Could not chdir to %s", opt_socket_path);
+
+	if (unlink(notify_sock_fullpath) == -1 && errno != ENOENT)
+		pexitf("Failed to remove existing notify socket: %s", notify_sock_fullpath);
+
+	if (bind(notify_socket_fd, (struct sockaddr *)&notify_addr, sizeof(struct sockaddr_un)) == -1)
+		pexitf("Failed to bind notify socket: %s", notify_sock_fullpath);
+
+	if (chmod(notify_sock_fullpath, 0777))
+		pexit("Failed to change notify socket permissions");
+
+	if (chdir(cwd) == -1)
+		pexitf("Could not chdir to %s", cwd);
```

Comment:
same here, I am pretty sure there's a decent amount of duplication here

Reply:
Agreed. I think actually between attach and notify right now, the symlink is being removed, created, socket bound.. then the next one removes it, creates the symlink, binds the socket... So it's def duplicated.

I'd like your read on the approach here - I found with rootless that using "notify/notify.sock" as the socket name exceeded the unix socket path.  The nice thing is, since I hardcoded it to always be notify/notify.sock, I know exactly how long it's going to be ... The downside being that it will ALWAYS be notify.sock in the container... we could have some play with the target directory if we implement an ENV or a cmdline option for that.  But regardless in rootless it was exceeding the length of sun_path, so I chdir'd to the user_socket_directory so I could be sure the relative path from there was < sizeof(sun_path)

If that approach is solid, it should probably be done that way in both attach and notify  (console for some reason takes a different approach of just creating a socket in /tmp?)... For that matter symlinks could be removed entirely and we could use the bundledata path if we just use ./attach and ./notify/notify.sock instead of putting the whole path in there.

Reply:
ah I didn't notice the chdir....

I am unsure of the approach to be honest. If we were going to run the conmon in a specific dir, we should make clear in documentation that is where.

does just `notify.sock` exceed the length? why have the specific dir? The container directory passed should be mostly touched by conmon and its caller, so it doesn't necessarily have to be very pretty.

I think I'd prefer to not have the chdir and instead go through the linking shenannigans we currently do

Reply:
this is still unresolved, I am hesitant of the chdir

Reply:
Why are you hesitant? It chdirs back to CWD at the end...

Reply:
haircommander:cri-o:  10:25 AM
I've never used it in a c program and I am worried about suddenly relying on relative paths. also, if we exceed PATH_MAX, won't that fail in other ways and places other than conmon?
I guess ultimately it feels hacky (despite the other work around being similarly hacky)

Joseph Gooch  10:26 AM
It IS hacky, we’re working around unix socket path length issues :smile:

haircommander:cri-o:  10:27 AM
oh right lol

Joseph Gooch  10:27 AM
I chdir, and then chdir back at the end of the function
if we exceed PATH_MAX, I imagine chdir will fail (which is fatal)

haircommander:cri-o:  10:28 AM
are you opposed to just doing notify.sock without the notify directory?

Joseph Gooch  10:28 AM
Can’t
Podman has to bindmount a directory
To set selinux labels
Plus the bind mount happens at a different time the unix socket is created
So Podman exposes the folder, conmon owns the socket within the folder
Otherwise bindmounting the socket itself gets… problematic
And we wouldn’t want to bindmount our entire /var or /run podman state folder

haircommander:cri-o:  10:31 AM
okay I'm convinced. not super stoked about it, but convinced :slightly_smiling_face: